src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,292 @@
+
+lemma coe.add_monoid_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.add_monoid_hom R K) x v = x v := rfl
+
+/-- The inclusion of `R_hat` in `K_hat` is a ring homomorphism. -/
+noncomputable! def coe.ring_hom : ring_hom (R_hat R K) (K_hat R K)  :=
+{ to_fun   := coe,
+  map_one' := rfl,
+  map_mul' := λ x y, by {ext p, simp only [pi.mul_apply, subring.coe_mul], refl },
+  ..coe.add_monoid_hom R K }
+
+lemma coe.ring_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.ring_hom R K) x v = x v := rfl
+
+end prod_completions_integers
+
+section algebra_instances
+
+instance : algebra K (K_hat R K) :=
+(by apply_instance : algebra K ((Π (v : height_one_spectrum R), v.adic_completion K)))
+
+instance prod_completions.algebra' : algebra R (K_hat R K) :=
+(by apply_instance : algebra R (Π (v : height_one_spectrum R), v.adic_completion K))
+
+instance : is_scalar_tower R K (K_hat R K) :=
+(by apply_instance : is_scalar_tower R K ((Π (v : height_one_spectrum R), v.adic_completion K)))
+
+instance : algebra R (R_hat R K) :=
+(by apply_instance : algebra R (Π (v : height_one_spectrum R), v.adic_completion_integers K))
+
+instance prod_completions.algebra_completions : algebra (R_hat R K) (K_hat R K) :=
+(prod_completions_integers.coe.ring_hom R K).to_algebra
+
+instance prod_completions.is_scalar_tower_completions :
+  is_scalar_tower R (R_hat R K) (K_hat R K) :=
+(by apply_instance : is_scalar_tower R (Π (v : height_one_spectrum R), v.adic_completion_integers K)
+   (Π (v : height_one_spectrum R), v.adic_completion K))
+
+end algebra_instances
+
+namespace prod_completions_integers
+
+/-- The inclusion of `R_hat` in `K_hat` is a ring homomorphism. -/
+noncomputable! def coe.alg_hom : alg_hom R (R_hat R K) (K_hat R K)  :=
+{ to_fun    := coe,
+  commutes' := λ r, rfl,
+  ..coe.ring_hom R K  }
+
+lemma coe.alg_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.alg_hom R K) x v = x v := rfl
+
+end prod_completions_integers
+
+/-! ### The finite adèle ring of a Dedekind domain
+We define the finite adèle ring of `R` as the restricted product over all maximal ideals `v` of `R`
+of `adic_completion` with respect to `adic_completion_integers`. We prove that it is a commutative
+ring. TODO: show that it is a topological ring with the restricted product topology. -/
+
+/-- A tuple `(x_v)_v` is in the restricted product of the `adic_completion` with respect to
+`adic_completion_integers` if for all but finitely many `v`, `x_v ∈ adic_completion_integers`. -/
+def restricted : K_hat R K → Prop := λ x,
+ ∀ᶠ (v : height_one_spectrum R) in filter.cofinite, (x v ∈ v.adic_completion_integers K)
+
+/-- The finite adèle ring of `R` is the restricted product over all maximal ideals `v` of `R`
+of `adic_completion` with respect to `adic_completion_integers`.-/
+def finite_adele_ring := { x : (K_hat R K) //
+  ∀ᶠ (v : height_one_spectrum R) in filter.cofinite, (x v ∈ v.adic_completion_integers K) }
```

Comment:
I would expect it should be less work (and more useful) to define these as a `subring`:
```suggestion
noncomputable! def finite_adele_ring : subring (K_hat R K) :=
{ carrier := λ x, ∀ᶠ (v : height_one_spectrum R) in filter.cofinite, (x v ∈ v.adic_completion_integers K),
  mul_mem' := sorry,
  one_mem' := sorry,
  add_mem' := sorry,
  zero_mem' := sorry,
  neg_mem' := sorry, }
```
and silently rely on `set_like.has_coe_to_sort` and `subring.to_comm_ring`.

Did you try this?

Reply:
I chose not to define it as a subring to be able to use dot notation for lemmas involving terms of `finite_adele_ring` (e.g., `x.coe_add y`).

Reply:
I agree with @ocfnash here: I don't think dot notation is an especially strong argument, especially since with a `subring` you can use `coe_add x y` (aka `_root_.coe_add x y`)

Reply:
The other big advantage is that you get lemmas like `coe_pow` and `coe_finset_sum` for free.

Reply:
I have turned `finite_adele_ring` into a subring. @ocfnash, @eric-wieser, could you take a look? Thanks.